addons/mail/models/res_users.py

```diff
@@ -268,42 +268,46 @@ def _init_messaging(self):
+        activities_by_rec_id_by_module_by_model_name = {}
+        group_custom_info = {}
+        models = list(dict.fromkeys(activities.mapped('res_model')))
+        for model in models:
+            activities_by_rec_id_by_module_by_model_name[model], group_custom_info[model] = self.env[model]._sort_activities_by_record_by_module(
+                activities.filtered_domain([('res_model', '=', model)]))
+        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_rec_id_by_module_by_model_name})
+        user_activities = []
+        for model_name, activities_by_rec_id_by_module in activities_by_rec_id_by_module_by_model_name.items():
+            domain = [("id", "in", list(set.union(*[{id for id in activities_by_rec_id} for activities_by_rec_id in activities_by_rec_id_by_module.values()])))]
+            allowed_records_ids = self.env[model_name].search(domain).ids
+            if not allowed_records_ids:
                 continue
-            module = self.env[model_name]._original_module
-            icon = module and modules.module.get_module_icon(module)
-            model = self.env["ir.model"]._get(model_name).with_prefetch(model_ids)
-            user_activities[model_name] = {
-                "id": model.id,
-                "name": model.name,
-                "model": model_name,
-                "type": "activity",
-                "icon": icon,
-                "total_count": 0,
-                "today_count": 0,
-                "overdue_count": 0,
-                "planned_count": 0,
-                "actions": [
-                    {
-                        "icon": "fa-clock-o",
-                        "name": "Summary",
-                    }
-                ],
-            }
-            for record, activities in activities_by_record.items():
-                if record not in allowed_records:
+            for module, activities_by_rec_id in activities_by_rec_id_by_module.items():
+                count = {'today': 0, 'overdue': 0, 'planned': 0}
+                for rec_id, activities in activities_by_rec_id.items():
+                    if rec_id not in allowed_records_ids:
+                        continue
+                    for activity in activities:
+                        count[activity.state] += 1
+                if not (count['today'] or count['planned'] or count['overdue']):
                     continue
-                for activity in activities:
-                    user_activities[model_name]["%s_count" % activity.state] += 1
-                    if activity.state in ("today", "overdue"):
-                        user_activities[model_name]["total_count"] += 1
-        return list(user_activities.values())
+                icon = module and modules.module.get_module_icon(module)
+                model = self.env["ir.model"]._get(model_name).with_prefetch(model_ids)
+                user_activities.append({
+                    "id": len(user_activities),
```

Comment:
Is this only used as a `t-key` for owl? If so it could be removed and the templates could use `t-as="group" t-key="group_index"` instead

Reply:
> Is this only used as a `t-key` for owl? If so it could be removed and the templates could use `t-as="group" t-key="group_index"` instead

Thanks for the feedback! This part was indeed not really good, I've changed that by:
- Renaming 'id' key in 'sequence' and only use it to sort activity groups in the systray (as it is the case now)
- Use group_index for the t-key as sequence can be duplicated for groups related to the same model

I've also added a test for that.